File: LeetCode/Easy/0257-binary-tree-paths/0257-binary-tree-paths.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public IList<string> BinaryTreePaths(TreeNode root) {
        var results = new List<string>();
        var path = new List<int>();
        
        TreeNode lastVisited = null;
        var current = root;
        var stack = new Stack<TreeNode>();

        while (current != null || stack.Count > 0)
        {
            while (current != null)
            {
                stack.Push(current);
                path.Add(current.val);
                current = current.left;
            }

            var peeked = stack.Peek();
            if (peeked.right != null && peeked.right != lastVisited)
            {
                current = peeked.right;
            }
            else
            {
                lastVisited = peeked;
                if (peeked.left == null && peeked.right == null)
                {
                    results.Add(string.Join("->", path));
                }
                stack.Pop();

                path.RemoveAt(path.Count - 1);
            }
        }

        return results;
    }
}

File: LeetCode/Easy/0257-binary-tree-paths/0257-binary-tree-paths.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:
        buffer = []
        current = root
        stack = []
        last_visited = None
        root_to_leaf_paths = []

        while current or stack:
            while current:
                buffer.append(str(current.val))
                stack.append(current)
                current = current.left

            node = stack[-1]
            if node.right and node.right is not last_visited:
                current = node.right
            else:
                stack.pop()
                if not node.left and not node.right:
                    root_to_leaf_paths.append("->".join(buffer))
                last_visited = node
                buffer.pop()
        
        return root_to_leaf_paths